server/src/internalClusterTest/java/org/opensearch/wlm/WorkloadManagementIT.java

```diff
@@ -0,0 +1,403 @@
+                    listener.onResponse(new TestResponse());
+                }
+            });
+        }
+    }
+
+    public static class TestQueryGroupTaskRequest<T extends Task> extends ActionRequest {
+        private final String type;
+        private final String queryGroupId;
+        private TaskFactory<T> taskFactory;
+
+        public TestQueryGroupTaskRequest(String type, String queryGroupId, TaskFactory<T> taskFactory) {
+            this.type = type;
+            this.queryGroupId = queryGroupId;
+            this.taskFactory = taskFactory;
+        }
+
+        public TestQueryGroupTaskRequest(StreamInput in) throws IOException {
+            super(in);
+            this.type = in.readString();
+            this.queryGroupId = in.readString();
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
+        }
+
+        @Override
+        public Task createTask(long id, String type, String action, TaskId parentTaskId, Map<String, String> headers) {
+            return taskFactory.createTask(id, type, action, "", parentTaskId, headers);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeString(type);
+            out.writeString(queryGroupId);
+        }
+
+        public String getType() {
+            return type;
+        }
+
+        public String getQueryGroupId() {
+            return queryGroupId;
+        }
+    }
+
+    public static class TestQueryGroupTaskTransportAction extends HandledTransportAction<TestQueryGroupTaskRequest, TestResponse> {
+        public static final ActionType<TestResponse> ACTION = new ActionType<>("internal::test_query_group_task_action", TestResponse::new);
+        private final ThreadPool threadPool;
+
+        @Inject
+        public TestQueryGroupTaskTransportAction(TransportService transportService, ThreadPool threadPool, ActionFilters actionFilters) {
+            super(ACTION.name(), transportService, actionFilters, TestQueryGroupTaskRequest::new);
+            this.threadPool = threadPool;
+        }
+
+        @Override
+        protected void doExecute(Task task, TestQueryGroupTaskRequest request, ActionListener<TestResponse> listener) {
+            threadPool.getThreadContext().putHeader(QUERY_GROUP_ID_HEADER, request.getQueryGroupId());
+            threadPool.executor(ThreadPool.Names.SEARCH).execute(() -> {
+                try {
+                    CancellableTask cancellableTask = (CancellableTask) task;
+                    ((QueryGroupTask) task).setQueryGroupId(threadPool.getThreadContext());
+                    assertEquals(request.getQueryGroupId(), ((QueryGroupTask) task).getQueryGroupId());
+                    long startTime = System.nanoTime();
+
+                    while (System.nanoTime() - startTime < TIMEOUT.getNanos()) {
```

Comment:
Do we really need to spend 10 seconds in completing this request to make it resource intensive ? Is there a more definite and clear way of ensuring the resource usage for the request ?

How about just running it more than required time ? how about just making it run for one second since the cpu usage is calculated using following
```
% CPU usage = (startTime - finishTime) / (totalTimeOnCpu)
```

here there is no reason for this task to get preempted (No I/O(network or disk) call, no voluntary sleep) hence should continuously be running .

Reply:
We're breaking out the 10 second while loop immediately when the task is cancelled. So if the task should be cancelled (wlm mode enabled + enforced mode), it will take a very short time in the loop. But I'll see if setting it to 1s work to reduce the time